### PR TITLE
PromQL: Add trigonometric functions

### PR DESCRIPTION
**Pull request by LeviHarrison**
This PR adds the following functions:
- `sin()`
- `cos()`
- `tan()` 
- `asin()`
- `acos()` 
- `atan()` 
- `sinh()`
- `cosh()`
- `tanh()`
- `asinh()`
- `acosh()`
- `atanh()`
- `deg()` 
- `rad()`
- `pi()`

Although `deg()` and `rad()` aren't strictly needed; their functions can be done in a more expanded form using `pi()`, I thought it'd be nice to have them anyway given how common their use is with trigonometry.

Fulfills #9233

### CODE REVIEW

File: docs/querying/functions.md [hunk 1]
```diff
@@ -434,3 +434,45 @@ over time and return an instant vector with per-series aggregation results:
...
+## Trigonometric Functions
+
+- `acos(v instant-vector)`: calculates the arccosine, in radians, of all elements in `v`. Special cases are:
+    - `acos(x) = NaN if x < -1 or x > 1`
+- `asin(v instant-vector)`: calculates the arcsine, in radians, of all elements in `v`. Special cases are:
+    - `asin(±0) = ±0`
+- `atan(v instant-vector)`: calculates the arctangent, in radians, of all elements in `v`. Special cases are:
+    - `atan(±0) = ±0`
+    - `atan(±Inf) = ±Pi/2`
+- `atan2(y, x instant-vector)`: calculates the arctangent of `y`/`x` for all elements, in radians, using the signs of the two to determine the quadrant of the result ([from the package](https://pkg.go.dev/math#Atan2)). Special cases are:
```
Comment (LeviHarrison) [inline docs/querying/functions.md:447]: The best explanation I could come up with is mostly lifted from the `math` docs, so I included a link to it.

**Comment by roidelapluie** [comment]
Can we: 1. specify that all functions use rad and 2. link to go math package for the special cases? I do not think we should document them ourselves.

File: promql/functions.go [hunk 2]
```diff
@@ -570,6 +570,70 @@ func funcLog10(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper
...
+	return simpleFunc(vals, enh, math.Acos)
+}
+
+// == atan(Vector parser.ValueTypeVector) Vector ===
+func funcAtan(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) Vector {
+	return simpleFunc(vals, enh, math.Atan)
+}
+
+// === atan2(Vector1, Vector2 parser.ValueTypeVector) Vector ===
+func funcAtan2(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) Vector {
```
Comment (roidelapluie) [inline promql/functions.go:604]: I do not think we ever have a function that takes 2 vectors.

**Comment by roidelapluie** [comment]
Could you also add the functions to the UI autocomplete? (pending https://github.com/prometheus/prometheus/pull/9229)

File: promql/parser/functions.go [hunk 3]
```diff
@@ -201,16 +226,32 @@ var Functions = map[string]*Function{
 		Variadic:   1,
 		ReturnType: ValueTypeVector,
 	},
+	"pi": {
+		Name:       "pi",
+		ArgTypes:   []ValueType{},
+		Variadic:   1,
```
Comment (roidelapluie) [inline promql/parser/functions.go:232]: Pi() does not taje an optional parameter.

File: promql/functions.go (hunk 2 above)
Comment (LeviHarrison) [reply promql/functions.go:604]: Yes, but there's not really any other way to implement `atan2`.

In the example provided by @leonerd:

> Having done that, your longerterm averages over data_x and data_y now make sense, and you can then extract an "average" magnitude and direction out of these again:
```
data_r:avg1d = sqrt(data_x:avg1d ^ 2 + data_y:avg1d ^ 2)
data_theta:avg1d = atan2(data_y:avg1d, data_x:avg1d)
```

I don't think it's meant to use the UNIX time and as the x.

**Comment by LeviHarrison** [comment]
> specify that all functions use rad

```
calculates the arccosine, in radians, of all elements in `v`.
```

> link to go math package for the special cases? I do not think we should document them ourselves.

Makes sense.

I do, do I not?

> Could you also add the functions to the UI autocomplete? (pending #9229)

Yeah, I will once it goes through.

**Comment by roidelapluie** [comment]
So, let's not have atan2 for the moment. For the math package, I mean we should not list all the exceptions ourselves.

**Comment by roidelapluie** [comment]
I would also split the trig functions in their own test file, WDYT?

**Comment by LeviHarrison** [comment]
> So, let's not have atan2 for the moment.

Is there a problem with having 2 vector arguments though?

> I would also split the trig functions in their own test file, WDYT?

Yeah, they're pretty long. I got bored of doing it so I wrote a generator 🤣

File: promql/parser/functions.go (hunk 3 above)
Comment (LeviHarrison) [reply promql/parser/functions.go:232]: Whoops didn't notice that when I copied it from another function.

**Comment by roidelapluie** [comment]
Yes there is a problem. Vector matching does not belong to PromQL functions in the current model. You have no guarantees that the vectors are matched.

**Comment by roidelapluie** [comment]
> > So, let's not have atan2 for the moment.
> 
> Is there a problem with having 2 vector arguments though?
> 
> > I would also split the trig functions in their own test file, WDYT?
> 
> Yeah, they're pretty long. I got bored of doing it so I wrote a generator

Since there is no logic on our side, we should only test one or two cases per function, with the addition of some specific tests for each (edge cases). I do not think we should do more, and because the special cases varies per function,  generating that does not seem easy.

**Comment by LeviHarrison** [comment]
Ok, I'll remove some of the normal tests and add edge cases.

**Comment by LeviHarrison** [comment]
There really is no logic on our side, so maybe even special cases don't matter, given that we don't influence their output. I think just one test per function making sure everything is hooked up properly works.

**Comment by roidelapluie** [comment]
> There really is no logic on our side, so maybe even special cases don't matter, given that we don't influence their output. I think just one test per function making sure everything is hooked up properly works.

I think it is still good to test that we handle NaN's properly.

**Comment by LeviHarrison** [comment]
I left the three tests for `rad()` and `deg()` because we implement the math ourselves.

**Comment by leonerd** [comment]
It's unfortunate that `atan2()` has to be left out for now. Will this be added in a later revision? It's the useful final function (along with `sin()` and `cos()`) for converting between polar and cartesian coordinate systems.

File: docs/querying/functions.md [hunk 4]
```diff
@@ -434,3 +434,15 @@ over time and return an instant vector with per-series aggregation results:
 
 Note that all values in the specified interval have the same weight in the
 aggregation even if the values are not equally spaced throughout the interval.
+
+## Trigonometric Functions
+
+- `acos(v instant-vector)`: calculates the arccosine, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Acos)).
+- `asin(v instant-vector)`: calculates the arcsine, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Asin)).
+- `atan(v instant-vector)`: calculates the arctangent, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Atan)).
+- `cos(v instant-vector)`: calculates the cosine, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Cos)).
```
Comment (leonerd) [inline docs/querying/functions.md:443]: I would say that it's the elements in `v` that are in radians here, not the value of the calculation. The inverse functions calculate the value "in radians", but the forward functions calculate a value of "all the elements in `v` in radians".

File: docs/querying/functions.md [hunk 5]
```diff
@@ -434,3 +434,15 @@ over time and return an instant vector with per-series aggregation results:
...
+- `deg(v instant-vector)`: converts radians to degrees for all elements in `v`.
+- `pi()`: returns the first 16 digits of Pi.
+- `rad(v instant-vector)`: converts degrees to radians for all elements in `v`.
+- `sin(v instant-vector)`: calculates the sine, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Sin)).
+- `tan(v instant-vector)`: calculates the tangent, in radians, of all elements in `v` ([special cases](https://pkg.go.dev/math#Tan)).
```
Comment (leonerd) [inline docs/querying/functions.md:448]: ditto these two

File: docs/querying/functions.md (hunk 4 above)
Comment (roidelapluie) [reply docs/querying/functions.md:443]: We can specify "in radians" as an introduction of the trig section.

Comment (leonerd) [reply docs/querying/functions.md:443]: Yes that's probably simpler, avoids some otherwise awkward wording.

Comment (LeviHarrison) [reply docs/querying/functions.md:443]:
```
All trigonometric operations are in radians.
```

WDYT? `rad()`, `deg()`, `pi()` aren't in radians so I don't think we can label all "functions" as in radians.

Comment (leonerd) [reply docs/querying/functions.md:443]: How about:

"The trigonometric functions work in radians:
[docs for arccos, arcsin, arctan, cos, sin, tan]

The following are useful for converting between degrees and radians:
[deg, rad, pi]"

Comment (roidelapluie) [reply docs/querying/functions.md:443]: That sounds great @leonerd

**Comment by leonerd** [comment]
From a user perspective, those docs look fine to me.

File: promql/parser/functions.go [hunk 6]
```diff
@@ -201,16 +221,31 @@ var Functions = map[string]*Function{
 		Variadic:   1,
 		ReturnType: ValueTypeVector,
 	},
+	"pi": {
+		Name:       "pi",
+		ArgTypes:   []ValueType{},
+		ReturnType: ValueTypeVector,
```
Comment (roidelapluie) [inline promql/parser/functions.go:227]: pi returns a scalar

Comment (LeviHarrison) [reply promql/parser/functions.go:227]: 🤦🏻‍♂️

**Comment by roidelapluie** [comment]
What would you think about explicit names for `rad` and `deg`? I can see how non-intuitive those names can be.

**Comment by LeviHarrison** [comment]
I think `rad` and `deg` are fine. From my experience, the abbreviation is very common in mathematical tools. For example, https://www.desmos.com/scientific.

**Comment by roidelapluie** [comment]
cc @bzed @juliusv your input is appreciated here.

**Comment by bzed** [comment]
Looks good to me, thanks a lot @LeviHarrison !

Regarding the naming of reg/deg: I think its fine, even my old school calculator uses that.

File: promql/testdata/trig_functions.test [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+eval instant at 5m tan(trig)
+	{l="x"} 0.6483608274590867
+	{l="y"} 2.2371609442247427
+	{l="NaN"} NaN
+
+eval instant at 5m asin(trig - 10.1)
+  {l="x"} -0.10016742116155944
+  {l="y"} NaN
+	{l="NaN"} NaN
```
Comment (juliusv) [inline promql/testdata/trig_functions.test:26]: Something ended up inconsistently indented here and in other tests below? :thinking:

**Comment by beorn7** [comment]
Now that we are on it, shouldn't we include hyperbolic functions, too?

**Comment by LeviHarrison** [comment]
If there's a use case, sure. I just held back because I didn't want to bloat.

**Comment by juliusv** [comment]
Looks all mostly good to me, but yeah, we should probably get https://github.com/prometheus/prometheus/pull/9229 merged first, so that we can add the new functions to the frontend in the same PR (lest we forget and release a semi-broken Prometheus again).

I have no opinion on hyperbolic functions. I had to look up what they were :)

File: promql/testdata/trig_functions.test (hunk 7 above)
Comment (LeviHarrison) [reply promql/testdata/trig_functions.test:26]: Whoops, we're doing tabs, not spaces.

**Comment by beorn7** [comment]
Not sure about the use case for hyperbolic functions. I would more argue we should have them to do the least surprising thing. (Let's imagine we have use cases for cos and sin but not for tan. We would probably still add tan, just for completeness. The hyperbolic functions are less obvious members of the family, but I see it in a similar area.)

**Comment by juliusv** [comment]
@leonerd 

> It's unfortunate that `atan2()` has to be left out for now. Will this be added in a later revision? It's the useful final function (along with `sin()` and `cos()`) for converting between polar and cartesian coordinate systems.

The problem with `atan2()` would be as @roidelapluie mentioned that we would now have to implement the concept of vector element matching (along with possibly grouping and matching options) that currently only exists for binary operators, in a normal PromQL function. Theoretically we could implement `atan2()` as a binary operator `foo atan2 bar`, but that would probably be strange? :)

**Comment by roidelapluie** [comment]
I am not against `foo atan2 bar`, but it would be different than our traditional binary matching. This is not the first time this function atan2 is requested.

**Comment by roidelapluie** [comment]
(therefore I would try out atan2 in another PR)

**Comment by leonerd** [comment]
@juliusv 

> The problem with `atan2()` would be as @roidelapluie mentioned that we would now have to implement the concept of vector element matching (along with possibly grouping and matching options) that currently only exists for binary operators, in a normal PromQL function. Theoretically we could implement `atan2()` as a binary operator `foo atan2 bar`, but that would probably be strange? :)

Oh, indeed I agree that would be strange, and yes it seems best to get the simple ones done first now. I'm just hoping the `atan2` one comes along sometime soon as well, because without that this only half-solves my problem.

But it's still a good start. :)

**Comment by LeviHarrison** [comment]
I've opened #9248 to explore the possibility of an `atan2` binary operator.

**Comment by LeviHarrison** [comment]
At this point, we may want to add the in inverse hyperbolic functions, as they seem to be the only common ones left.

**Comment by roidelapluie** [comment]
> At this point, we may want to add the inverse hyperbolic functions, as they seem to be the only common ones left.

Yes, let's get them so we have full support

**Comment by Nexucis** [comment]
#9229 is finished. It just need to be reviewed now and merged. It shouldn't block this PR too long now :).
It took a bit of time to figure out how to integrate the grammar directly in Prometheus, but it's done now.

**Comment by LeviHarrison** [comment]
Sounds tricky, thank you so much!

**Comment by juliusv** [comment]
#9229 is merged, so it should be possible to do all necessary (backend + frontend) changes in this PR now :)

**Comment by juliusv** [comment]
Thanks! Looks good to me. While the lezer PromQL grammar stuff for the frontend is now in the Prometheus repo, @Nexucis is still working on actually building from it, so this PR doesn't yet work fully for the frontend stuff. But I think we just have to wait for @Nexucis on that. I'd just still hold off a tiny bit here before merging to make sure that everything works for the new functions on the frontend side, after that integration is done.

**Comment by Nexucis** [comment]
lgtm too :). Sorry @LeviHarrison for blocking again your PR. I wasn't expected that :sweat_smile:

this PR is blocked by #9227 now.

Hopefully with @juliusv we are going to solve some issues we have when integrating locally the codemirror-promql in react.

**Comment by Nexucis** [comment]
hey @LeviHarrison,

Integration of `codemirror-promql` in the `react-app` is finally finished.
Could you solve the conflicts you have ? Hopefully, it shouldn't be too complicated to solve them.

I think then we will be able to merge this PR, since there is no other blocking point

**Comment by LeviHarrison** [comment]
I'll get to those conflicts as soon as I can, thanks for getting this sorted!

**Comment by LeviHarrison** [comment]
Should be resolved

**Comment by juliusv** [comment]
Amazing, looks good to me! :+1: Thanks for the patience :)

**Comment by LeviHarrison** [comment]
Thanks for everyone's time reviewing!

**Comment by LeviHarrison** [comment]
Is there any way we can maybe squash the commits after the merge because this was quite a few?

**Comment by roidelapluie** [comment]
> Is there any way we can maybe squash the commits after the merge? Because this was quite a few.

We are not force pushing main, so, no.

**Comment by juliusv** [comment]
Oh, I thought I did "Squash & Merge", which is the default for me normally, but I think GitHub still had "Create a merge commit" preselected for me this time because I had to use that recently in another PR.

Sorry about that, didn't want to create that much noise in `main` :(

**Pull request merged**